File: easy/array/mountain_peak/mountain_peak.py
from typing import List


class Solution:
    def peakIndexInMountainArray(self, A: List[int]) -> int:
        start, end = 0, len(A)
        while start < end:
            mid = int((start+end)/2)
            if A[mid-1] < A[mid] and A[mid] > A[mid+1]:
                return mid
            elif A[mid+1] > A[mid]:
                start = mid+1
            elif A[mid-1] > A[mid]:
                end = mid


s = Solution()
print("Solution 1 : ", s.peakIndexInMountainArray([0, 1, 0]))
print("Solution 2 : ", s.peakIndexInMountainArray([0, 2, 1, 0]))

File: easy/array/mountain_peak/mountain_peak.go
package main

import "fmt"

func peakIndexInMountainArray(A []int) int {
	start, end := 0, len(A)
	for start < end {
		mid := (start + end) / 2
		if A[mid-1] < A[mid] && A[mid] > A[mid+1] {
			return mid
		} else if A[mid] < A[mid+1] {
			start = mid + 1
		} else if A[mid-1] > A[mid] {
			end = mid
		}
	}
	return -1
}

func main() {
	fmt.Println("Solution 1 : ", peakIndexInMountainArray([]int{0, 1, 0}))
	fmt.Println("Solution 2 : ", peakIndexInMountainArray([]int{0, 2, 1, 0}))
}
